### PR TITLE
Removed some unnecessary codegen from async state machine.

### PR DESCRIPTION
**Pull request by VSadov**
I was looking through our async codegen and noticed that we are zeroing out the awaiter local after fetching its Result, often right before exiting the MoveNext method. 
We also allocate unnecessary temp while doing this.

This pattern would repeat for every `await` and just makes the state machine method bigger. It is generally not necessary to zero-out locals. 

Basically a very simple `await` like ` Console.WriteLine(await Somehthing())` 
will end with 

```cs
                ...   some code here to handle suspending/resuming if necessary ...

		int temp = taskAwaiter.GetResult();
                // the following erasing of taskAwaiter seems unnecessary.
                // taskAwaiter is just a local variable.
		taskAwaiter = default(TaskAwaiter<int>);   
		Console.WriteLine(temp);
```

Where it can be just:

```cs
                ...   some code here to handle suspending/resuming if necessary ...

		Console.WriteLine(taskAwaiter.GetResult());
```

**Risk**

The risk is low. 
Storing a value into a regular (unpinned) local variable without ever reading it is not observable. The change essentially removes a dead store.

The only possible case where this can go wrong is if we are running with conservative GC that does not track liveness of stack roots (not common) + when containing async method is active for a very long time - like spinning in a loop or blocked on an IO. 
If GC is to happen after we are done with the given awaiter, GC may still consider the awaiter as rooted.

This would be a problem for any ref-containing local that went out of scope without being zeroed out, which is very common.  
So, basically either this change does not cause a problem or we have bigger problems (conservative GC). 

**Performance impact**

We are removing a dead local store here (which JIT might also be able to remove), so perf impact should be none or positive (possibly very slightly).
As evident from the the test changes, IL does get smaller.

I will look for any async benchmarks to run here.

**Is this a regression from a previous update?**

No. This is technically not a bug and we emitted such code for quite some time.

We have numerous tests that verify the behavior and codegen of async code. Most of this change is actually updating the IL baselines.

### CODE REVIEW

**Comment by VSadov** [comment]
@dotnet/roslyn-compiler - a trivial change. Most diffs are IL diffs in tests.

**Comment by VSadov** [comment]
@jaredpar - I'd hope to put this in Master, but if it does not meet the bar I can rebase to the readonly refs.

**Comment by VSadov** [comment]
@stephentoub - just to check if this is is not intentional to get around some gotchas with JIT not tracking the lifetime of awaiter or something.  
I think it should - it is a trivial struct with one ref field and we should not bother cleaning local like that.

**Comment by stephentoub** [comment]
>  it is a trivial struct with one ref field 

That's true of TaskAwaiter, but not of every possible awaiter type... even the awaiter you get back from ConfigureAwait is (slightly) more than that.

That said, I don't know that it really matters. I expect it's important in many situations to null out the awaiter field in the state machine, but much less so for any local temp. Seems like a good simplification.

**Comment by VSadov** [comment]
@stephentoub - Thanks!  
Yes, we definitely null-out awaiter fields once no longer need their data.

**Comment by VSadov** [comment]
@dotnet/roslyn-compiler - ping.

**Comment by VSadov** [comment]
Turns out we do not have an easily available benchmark suite targeting aspects of async.

I have entered a tracking bug https://github.com/dotnet/roslyn/issues/18806 to make a common location for benchmarks and start collecting them there.

**Comment by VSadov** [comment]
RE: benchmarks

For this change I just wrote a microbenchmark specifically targeting an await of a completed task - very common scenario and also the scenario that we are trying to improve here.

The benchmark shows small, but consistent **2% throughput improvement**. It is not a lot, but this is also a very small change.
(x64 .NET FX  RyuJIT, my dev machine)

I would expect that with bigger awaiter structs (like for ValueTask), the difference might be larger, since zeroing that out might take longer.

The following is my benchmarking code.
I plan to turn it into a real benchmark as described in https://github.com/dotnet/roslyn/issues/18806
And add a scenario for ValueTask as well.

=====

```C#
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ConsoleApp52
{
    class Program
    {
        static void Main(string[] args)
        {
            for (;;)
            {
                int averagingIterations = 50;
                long millis = 0;

                for (int i = 0; i < averagingIterations; i++)
                {
                    var sw = Stopwatch.StartNew();
                    Test1(1000000000).Wait();
                    sw.Stop();
                    millis += sw.ElapsedMilliseconds;  
                }

                System.Console.WriteLine(millis / averagingIterations);
            }
        }

        static async Task<int> Test1(int times)
        {
            int val = 0;

            for(int i = 0; i < times; i++) 
            {
                val += await One; 
            }

            return 0;
        }

        static Task<int> One = Task.FromResult(1);
    }
}
```

**Comment by VSadov** [comment]
test this please

**Comment by stephentoub** [comment]
> consistent 2% throughput improvement. It is not a lot, but this is also a very small change.

Nice!

**Comment by gafter** [comment]
Have you done a comparison to the code produced by VS2015?

**Comment by VSadov** [comment]
test this please

just to see the state of the build

**Comment by VSadov** [comment]
@gafter - I compared to VS2016. 

We did not have a lot of changes to async since 2015. We emit slightly better switches when multiple awaits are inside Try blocks. We also cache "this" into a local if it is used in an async method (to avoid re-fetching it from the display).   
Either change would be orthogonal to this one though, and would need a different benchmark to be measurable.  
It would be a good idea to have benchmarks for those.

**Comment by VSadov** [comment]
@gafter - I confused 16 and 15. 
The 2015 was still using native compiler. It would be an interesting comparison as well

**Comment by VSadov** [comment]
@gafter - for the purpose of this change VS2015 produced same code as VS2016 or VS2017. 
We had major differences in capturing strategy and stack spilling, but zeroing-out of awaiter temps did not change. 

It seems we have just replicated the behavior of native compiler here. Perhaps there was a scenario where in the native codegen the awaiters were not always locals, or it was just an artifact of emit strategy.

**Comment by VSadov** [comment]
test this please

just to see the state of the build

**Comment by VSadov** [comment]
per email discussion this is approved for 15.3 as a low risk perf work by @MeiChin-Tsai
